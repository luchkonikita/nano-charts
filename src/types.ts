export type Settings = {
  cssClassName: string
  strokeColor: string
  gridColor: string
  labelFontSize: number
  maxLabelHeight: number
  maxLabelWidth: number
  minContainerHeight: number
  mousemoveTimeout: number
  overlayOverlap: number
  paddingH: number
  paddingV: number
  pointHoverRadius: number
  pointRadius: number
  resizeTimeout: number
  tooltipOffsetX: number
  xGridOverflowSize: number
  yGridOverflowSize: number
}

export type Range = {
  min: number
  max: number
}

export type ChartData = {
  labels: string[]
  samples: {
    [index: string]: number[]
  }
  options?: {
    // TODO: Write better type
    [index: string]: {
      stroke?: string
      fill?: string
    }
  }
}

export type ViewportCoordinate = [number, number]

export type Viewport = {
  height: number
  width: number
}

export type RangesConstraint = {
  yValues: Range,
  xValues: Range
}

export type Attributes = {
  [index: string]: string | number
}

export type TooltipRendererSample = {
  key: string
  value: string
  color: string
}

export type TooltipRenderer = (label: string, samples: TooltipRendererSample[]) => string
