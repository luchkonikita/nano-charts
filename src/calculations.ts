import {
  ChartData,
  Range,
  RangesConstraint
} from './types'

export function getValuesForRange(range: Range): number[] {
  const diff = range.max - range.min

  // Don't bother with prettifying too small numbers
  if (diff < 1) return [range.min, range.max]

  // Define default step which is diff divided by 10
  let steps = 2
  let step = diff / steps

  // Dividers considered nice, values like 0, 5, 10, 15 are usually looking good
  const dividers = [2, 5]

  // Try applying them all, splitting in smaller parts will win
  dividers.forEach(divider => {
    for (let index = 1; index <= 10; index++) {
      const maybeBetterStep = diff / index
      const isDividable = maybeBetterStep % divider === 0

      if (isDividable) {
        steps = index
        step = maybeBetterStep
      }
    }
  })

  let current = range.min
  const result = [range.min]

  // We need to fill the range between min and max values
  for (let index = 0; index < (steps - 1); index++) {
    current += step
    result.push(Math.floor(current))
  }
  result.push(range.max)

  return result
}

function extendNumber(n: number, round: 'floor' | 'ceil'): number {
  if (n === 0) return 0
  const str = Math.floor(Math.abs(n)).toString()
  const factor = n === 0 ? 0 : n / Math.abs(n)
  let firstNum = parseInt(str[0], 10) * factor
  if (round === 'ceil') {
    firstNum = firstNum + 1
  } else {
    firstNum = firstNum - 1
  }
  return firstNum * Math.pow(10, str.length - 1)
}

export function extendRange(range: Range): Range {
  return {
    min: extendNumber(range.min, 'floor'),
    max: extendNumber(range.max, 'ceil')
  }
}

export function getRangesConstraint(data: ChartData): RangesConstraint  {
  const firstKey = Object.keys(data.samples)[0]

  // Get a real sample so the chart lines will fill the container
  let minYData: number = data.samples[firstKey][0]
  let maxYData: number = data.samples[firstKey][0]
  const minXData = 0
  const maxXData = data.labels.length - 1

  // Calculate min and max values
  Object.keys(data.samples).forEach(key => {
    const samples = data.samples[key]
    samples.forEach(sample => {
      if (sample > maxYData) maxYData = sample
      if (sample < minYData) minYData = sample
    })
  })

  return {
    yValues: extendRange({ min: minYData, max: maxYData }),
    xValues: { min: minXData, max: maxXData }
  }
}

export function calculateRangeProjection(baseRange: Range, targetRange: Range, value: number): number {
  if (value < baseRange.min || value > baseRange.max) {
    throw new Error('Value is not within the range')
  }
  if (baseRange.max < baseRange.min || targetRange.max < targetRange.min) {
    throw new Error('Invalid range range')
  }

  if (value === baseRange.min) return targetRange.min
  if (value === baseRange.max) return targetRange.max

  const baseRangeTotal = baseRange.max - baseRange.min
  const targetRangeTotal = targetRange.max - targetRange.min

  const baseDelta = Math.abs(
    value < 0
      ? value - baseRange.min
      : value - Math.abs(baseRange.min)
  )
  const coefficient = baseDelta / baseRangeTotal
  return targetRange.min + (targetRangeTotal * coefficient)
}
