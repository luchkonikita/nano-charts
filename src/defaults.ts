import { Settings } from './types'

const DEFAULTS: Settings = {
  cssClassName: 'nano-charts',
  strokeColor: '#6a6a6a',
  gridColor: '#ececec',
  labelFontSize: 12,
  maxLabelHeight: 20,
  maxLabelWidth: 60,
  mousemoveTimeout: 50,
  overlayOverlap: 10,
  paddingH: 80,
  paddingV: 40,
  pointHoverRadius: 4,
  pointRadius: 2,
  resizeTimeout: 300,
  tooltipOffsetX: 10,
  xGridOverflowSize: 20,
  yGridOverflowSize: 5
}

export default DEFAULTS
