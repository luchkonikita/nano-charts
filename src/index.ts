import Chart from './chart'
import { ChartData } from './types'

export { ChartData }
export default Chart
