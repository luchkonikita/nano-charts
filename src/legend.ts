import { Settings } from './types'
import { createElement } from './utils'

export default class Legend {
  public el: HTMLDivElement

  constructor(private settings: Settings) {
    this.el = createElement('div', {class: settings.cssClassName + '-legend'}) as HTMLDivElement
  }

  public update(datasets: Array<{key: string, color: string}>) {
    const {settings} = this
    let content = ''

    datasets.forEach(({key, color}) => {
      const badgeStyle = `background-color: ${color}; color: ${color};`
      content += `
        <span class='${settings.cssClassName + '-legend-item'}'>
          <span class='${settings.cssClassName + '-legend-item-badge'}' style='${badgeStyle}'></span>
          ${key}
        </span>
      `
    })

    this.el.innerHTML = content
  }
}
