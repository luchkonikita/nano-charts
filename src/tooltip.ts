import { Settings, TooltipRendererSample } from './types'
import { createElement } from './utils'

export default class Tooltip {
  public el: HTMLDivElement

  constructor(private settings: Settings) {
    this.el = createElement('div', {class: settings.cssClassName + '-tooltip'}) as HTMLDivElement
  }

  public update(label: string, samples: TooltipRendererSample[]) {
    const { cssClassName } = this.settings
    let content = `<header class="${cssClassName}-title">${label}</header>`
    samples.forEach(({color, key, value}) => {
      content += `
        <p class="${cssClassName}-item">
          <span class="${cssClassName}-badge" style="background-color: ${color};"></span>
          ${key}: ${value}
        </p>
      `
    })
    this.el.innerHTML = content
  }

  public show() {
    this.el.classList.add('is-visible')
  }

  public hide() {
    this.el.classList.remove('is-visible')
  }

  public moveTo(x: number, y: number) {
    this.el.style.transform = `translate(${x}px, ${y}px)`
  }

  public getBoundingClientRect() {
    return this.el.getBoundingClientRect()
  }
}
