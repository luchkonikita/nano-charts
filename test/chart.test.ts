import Chart from '../src/chart'
import { ChartData } from '../src/types'
import { createMockDiv, resizeMockDiv } from './helpers'

const chartData: ChartData = {
  labels: ['January', 'February', 'March'],
  samples: {
    'Metric A': [0, 5, 10],
    'Metric B': [-10, 5, 20]
  },
  options: {
    'Metric A': {
      color: 'violet'
    }
  }
}

describe('Chart', () => {
  let element: HTMLElement
  let chart: Chart

  beforeEach(() => {
    // HACK: jsdom does not implement this, so we need to mimic the browser behaviour.
    // Let's say each character spans 5 pixels regardless of font settings and other circumstances.
    Object.defineProperty(SVGElement.prototype, 'textLength', {
      configurable: true,
      get() {
        const imaginaryCharWidth = 10
        const charCount = this.textContent.length
        return {baseVal: {value: charCount * imaginaryCharWidth}}
      }
    })

    window.document.body.innerHTML = ''
    element = createMockDiv(1000, 200)
    window.document.body.appendChild(element)
    chart = new Chart(element)
  })

  afterEach(() => {
    chart.destroy()
  })

  describe('initialization', () => {
    it('appends the SVG', () => {
      expect(element).toMatchSnapshot()
    })
  })

  describe('setData', () => {
    it('sets the data', () => {
      chart.setData(chartData)
      expect(element).toMatchSnapshot()
    })
  })

  describe('destroy', () => {
    it('cleans up the target node', () => {
      chart.destroy()
      expect(element).toMatchSnapshot()
    })
  })

  describe('resize', () => {
    it('re-calculates the coordinates', done => {
      chart.setData(chartData)
      resizeMockDiv(element as HTMLDivElement, 800, 200)
      window.dispatchEvent(new Event('resize'))

      setTimeout(() => {
        expect(element).toMatchSnapshot()
        done()
      }, 400)
    })
  })

  describe('mouseenter and mouseleave', () => {
    it('sets flags', () => {
      chart.setData(chartData)
      const overlay = element.querySelector('.nano-charts-overlay') as HTMLElement
      const tooltip = element.querySelector('.nano-charts-tooltip') as HTMLElement

      expect(tooltip.className).not.toContain('is-visible')

      overlay.dispatchEvent(new Event('mouseenter'))
      expect(tooltip.className).toContain('is-visible')

      overlay.dispatchEvent(new Event('mouseleave'))
      expect(tooltip.className).not.toContain('is-visible')
    })
  })

  describe('mousemove', () => {
    it('highlights the metrics in the hovered are', async () => {
      chart.setData(chartData)
      const overlay = element.querySelector('.nano-charts-overlay') as HTMLElement

      async function assertRadiuses(asserted: string[]) {
        return new Promise(resolve => {
          setTimeout(() => {
            const circles = element.querySelectorAll('circle')
            const radiuses = Array.prototype.map.call(circles, (circle: SVGCircleElement) => circle.getAttribute('r'))
            expect(radiuses).toStrictEqual(asserted)
            resolve()
          }, 400)
        })
      }

      overlay.dispatchEvent(new MouseEvent('mouseenter'))
      overlay.dispatchEvent(new MouseEvent('mousemove', { clientX: 0 }))
      await assertRadiuses(['4', '2', '2', '4', '2', '2'])

      overlay.dispatchEvent(new MouseEvent('mousemove', { clientX: 500 }))
      await assertRadiuses(['2', '4', '2', '2', '4', '2'])

      overlay.dispatchEvent(new MouseEvent('mousemove', { clientX: 1000 }))
      await assertRadiuses(['2', '2', '4', '2', '2', '4'])
    })

    it.skip('renders the tooltip with relevant data', () => {
      // Implement this...
    })
  })
})
