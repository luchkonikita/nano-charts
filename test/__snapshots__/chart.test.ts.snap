// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart destroy cleans up the target node 1`] = `
<div
  style="height: 250px; position: relative;"
>
  <div
    class="nano-charts-legend"
  />
</div>
`;

exports[`Chart initialization allows passing data right away 1`] = `
<div
  style="height: 250px; position: relative;"
>
  <svg
    class="nano-charts"
    height="200"
    viewBox="0 0 1000 200"
    width="1000"
  >
    <g
      transform="translate(80, 40)"
    >
      <line
        stroke="#ececec"
        x1="0"
        x2="0"
        y1="-5"
        y2="125"
      />
      <line
        stroke="#ececec"
        x1="420"
        x2="420"
        y1="-5"
        y2="125"
      />
      <line
        stroke="#ececec"
        x1="840"
        x2="840"
        y1="-5"
        y2="125"
      />
      <line
        stroke="#ececec"
        x1="-20"
        x2="860"
        y1="120"
        y2="120"
      />
      <line
        stroke="#ececec"
        x1="-20"
        x2="860"
        y1="96"
        y2="96"
      />
      <line
        stroke="#ececec"
        x1="-20"
        x2="860"
        y1="72"
        y2="72"
      />
      <line
        stroke="#ececec"
        x1="-20"
        x2="860"
        y1="48"
        y2="48"
      />
      <line
        stroke="#ececec"
        x1="-20"
        x2="860"
        y1="24"
        y2="24"
      />
      <line
        stroke="#ececec"
        x1="-20"
        x2="860"
        y1="0"
        y2="0"
      />
      <path
        d="M 0 72 L 420 84 L 840 96"
        fill="transparent"
        stroke="violet"
        stroke-width="1"
      />
      <circle
        cx="0"
        cy="72"
        fill="violet"
        r="2"
      />
      <circle
        cx="420"
        cy="84"
        fill="violet"
        r="2"
      />
      <circle
        cx="840"
        cy="96"
        fill="violet"
        r="2"
      />
      <path
        d="M 0 72 L 420 84 L 840 96 L 840 72 L 0 72"
        fill="violet"
        stroke="transparent"
        stroke-width="1"
      />
      <path
        d="M 0 96 L 420 84 L 840 120"
        fill="transparent"
        stroke="#6a6a6a"
        stroke-width="1"
      />
      <circle
        cx="0"
        cy="96"
        fill="#6a6a6a"
        r="2"
      />
      <circle
        cx="420"
        cy="84"
        fill="#6a6a6a"
        r="2"
      />
      <circle
        cx="840"
        cy="120"
        fill="#6a6a6a"
        r="2"
      />
    </g>
    <g>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="middle"
        x="80"
        y="180"
      >
        Jan...
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="middle"
        x="500"
        y="180"
      >
        Feb...
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="middle"
        x="920"
        y="180"
      >
        March
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="end"
        x="40"
        y="160"
      >
        -20
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="end"
        x="40"
        y="136"
      >
        -10
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="end"
        x="40"
        y="112"
      >
        0
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="end"
        x="40"
        y="88"
      >
        10
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="end"
        x="40"
        y="64"
      >
        20
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="end"
        x="40"
        y="40"
      >
        30
      </text>
    </g>
  </svg>
  <div
    class="nano-charts-tooltip"
  />
  <div
    class="nano-charts-overlay"
  />
  <div
    class="nano-charts-legend"
  >
    
        
    <span
      class="nano-charts-legend-item"
    >
      
          
      <span
        class="nano-charts-legend-item-badge"
        style="background-color: violet; color: violet;"
      />
      
          Metric A
        
    </span>
    
      
        
    <span
      class="nano-charts-legend-item"
    >
      
          
      <span
        class="nano-charts-legend-item-badge"
        style="background-color: #6a6a6a; color: #6a6a6a;"
      />
      
          Metric B
        
    </span>
    
      
  </div>
  <style>
    
    .nano-charts {
      font-family:Arial, Helvetica, sans-serif;
    }
    .nano-charts-label {
      line-height: 20;
    }
    .nano-charts-legend {
      background-color: #f9f9f9;
      display: flex;
      font-family:Arial, Helvetica, sans-serif;
      font-size: 12px;
      flex-wrap: wrap;
      justify-content: center;
      padding: 10px;
    }
    .nano-charts-legend-item {
      margin: 0 10px;
    }
    .nano-charts-legend-item-badge {
      border-radius: 4px;
      display: inline-block;
      height: 8px;
      margin-right: 6px;
      position: relative;
      width: 8px;
    }
    .nano-charts-legend-item-badge:before {
      background-color: currentColor;
      content: '';
      display: block;
      height: 2px;
      left: -3px;
      position: absolute;
      top: 3px;
      width: 14px;
    }
    .nano-charts-tooltip {
      background-color: rgba(0, 0, 0, 0.8);
      border: 1px solid #000;
      border-radius: 2px;
      color: #fafafa;
      font-family:Arial, Helvetica, sans-serif;
      font-size: 12px;
      left: 0;
      line-height: 16px;
      opacity: 0;
      padding: 16px;
      pointer-events: none;
      position: absolute;
      top: 0;
    }
    .nano-charts-tooltip.is-visible {
      opacity: 1;
    }
    .nano-charts-tooltip.is-visible:empty {
      opacity: 0;
    }
    .nano-charts-overlay {
      bottom: 30px;
      left: 70px;
      position: absolute;
      right: 70px;
      top: 30px;
    }
    .nano-charts-title {
      font-weight: 600;
      margin: 0 0 8px 0;
    }
    .nano-charts-item {
      margin: 0;
    }
    .nano-charts-badge {
      border-radius: 4px;
      display: inline-block;
      height: 8px;
      margin-right: 4px;
      width: 8px;
    }
  
  </style>
</div>
`;

exports[`Chart initialization appends the SVG 1`] = `
<div
  style="height: 250px; position: relative;"
>
  <svg
    class="nano-charts"
    height="200"
    viewBox="0 0 1000 200"
    width="1000"
  >
    <g
      transform="translate(80, 40)"
    />
    <g />
  </svg>
  <div
    class="nano-charts-tooltip"
  />
  <div
    class="nano-charts-overlay"
  />
  <div
    class="nano-charts-legend"
  />
  <style>
    
    .nano-charts {
      font-family:Arial, Helvetica, sans-serif;
    }
    .nano-charts-label {
      line-height: 20;
    }
    .nano-charts-legend {
      background-color: #f9f9f9;
      display: flex;
      font-family:Arial, Helvetica, sans-serif;
      font-size: 12px;
      flex-wrap: wrap;
      justify-content: center;
      padding: 10px;
    }
    .nano-charts-legend-item {
      margin: 0 10px;
    }
    .nano-charts-legend-item-badge {
      border-radius: 4px;
      display: inline-block;
      height: 8px;
      margin-right: 6px;
      position: relative;
      width: 8px;
    }
    .nano-charts-legend-item-badge:before {
      background-color: currentColor;
      content: '';
      display: block;
      height: 2px;
      left: -3px;
      position: absolute;
      top: 3px;
      width: 14px;
    }
    .nano-charts-tooltip {
      background-color: rgba(0, 0, 0, 0.8);
      border: 1px solid #000;
      border-radius: 2px;
      color: #fafafa;
      font-family:Arial, Helvetica, sans-serif;
      font-size: 12px;
      left: 0;
      line-height: 16px;
      opacity: 0;
      padding: 16px;
      pointer-events: none;
      position: absolute;
      top: 0;
    }
    .nano-charts-tooltip.is-visible {
      opacity: 1;
    }
    .nano-charts-tooltip.is-visible:empty {
      opacity: 0;
    }
    .nano-charts-overlay {
      bottom: 30px;
      left: 70px;
      position: absolute;
      right: 70px;
      top: 30px;
    }
    .nano-charts-title {
      font-weight: 600;
      margin: 0 0 8px 0;
    }
    .nano-charts-item {
      margin: 0;
    }
    .nano-charts-badge {
      border-radius: 4px;
      display: inline-block;
      height: 8px;
      margin-right: 4px;
      width: 8px;
    }
  
  </style>
</div>
`;

exports[`Chart resize re-calculates the coordinates 1`] = `
<div
  style="height: 200px; position: relative; width: 800px;"
>
  <svg
    class="nano-charts"
    height="200"
    viewBox="0 0 800 200"
    width="800"
  >
    <g
      transform="translate(80, 40)"
    >
      <line
        stroke="#ececec"
        x1="0"
        x2="0"
        y1="-5"
        y2="125"
      />
      <line
        stroke="#ececec"
        x1="320"
        x2="320"
        y1="-5"
        y2="125"
      />
      <line
        stroke="#ececec"
        x1="640"
        x2="640"
        y1="-5"
        y2="125"
      />
      <line
        stroke="#ececec"
        x1="-20"
        x2="660"
        y1="120"
        y2="120"
      />
      <line
        stroke="#ececec"
        x1="-20"
        x2="660"
        y1="96"
        y2="96"
      />
      <line
        stroke="#ececec"
        x1="-20"
        x2="660"
        y1="72"
        y2="72"
      />
      <line
        stroke="#ececec"
        x1="-20"
        x2="660"
        y1="48"
        y2="48"
      />
      <line
        stroke="#ececec"
        x1="-20"
        x2="660"
        y1="24"
        y2="24"
      />
      <line
        stroke="#ececec"
        x1="-20"
        x2="660"
        y1="0"
        y2="0"
      />
      <path
        d="M 0 72 L 320 84 L 640 96"
        fill="transparent"
        stroke="violet"
        stroke-width="1"
      />
      <circle
        cx="0"
        cy="72"
        fill="violet"
        r="2"
      />
      <circle
        cx="320"
        cy="84"
        fill="violet"
        r="2"
      />
      <circle
        cx="640"
        cy="96"
        fill="violet"
        r="2"
      />
      <path
        d="M 0 72 L 320 84 L 640 96 L 640 72 L 0 72"
        fill="violet"
        stroke="transparent"
        stroke-width="1"
      />
      <path
        d="M 0 96 L 320 84 L 640 120"
        fill="transparent"
        stroke="#6a6a6a"
        stroke-width="1"
      />
      <circle
        cx="0"
        cy="96"
        fill="#6a6a6a"
        r="2"
      />
      <circle
        cx="320"
        cy="84"
        fill="#6a6a6a"
        r="2"
      />
      <circle
        cx="640"
        cy="120"
        fill="#6a6a6a"
        r="2"
      />
    </g>
    <g>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="middle"
        x="80"
        y="180"
      >
        Jan...
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="middle"
        x="400"
        y="180"
      >
        Feb...
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="middle"
        x="720"
        y="180"
      >
        March
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="end"
        x="40"
        y="160"
      >
        -20
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="end"
        x="40"
        y="136"
      >
        -10
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="end"
        x="40"
        y="112"
      >
        0
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="end"
        x="40"
        y="88"
      >
        10
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="end"
        x="40"
        y="64"
      >
        20
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="end"
        x="40"
        y="40"
      >
        30
      </text>
    </g>
  </svg>
  <div
    class="nano-charts-tooltip"
  />
  <div
    class="nano-charts-overlay"
  />
  <div
    class="nano-charts-legend"
  >
    
        
    <span
      class="nano-charts-legend-item"
    >
      
          
      <span
        class="nano-charts-legend-item-badge"
        style="background-color: violet; color: violet;"
      />
      
          Metric A
        
    </span>
    
      
        
    <span
      class="nano-charts-legend-item"
    >
      
          
      <span
        class="nano-charts-legend-item-badge"
        style="background-color: #6a6a6a; color: #6a6a6a;"
      />
      
          Metric B
        
    </span>
    
      
  </div>
  <style>
    
    .nano-charts {
      font-family:Arial, Helvetica, sans-serif;
    }
    .nano-charts-label {
      line-height: 20;
    }
    .nano-charts-legend {
      background-color: #f9f9f9;
      display: flex;
      font-family:Arial, Helvetica, sans-serif;
      font-size: 12px;
      flex-wrap: wrap;
      justify-content: center;
      padding: 10px;
    }
    .nano-charts-legend-item {
      margin: 0 10px;
    }
    .nano-charts-legend-item-badge {
      border-radius: 4px;
      display: inline-block;
      height: 8px;
      margin-right: 6px;
      position: relative;
      width: 8px;
    }
    .nano-charts-legend-item-badge:before {
      background-color: currentColor;
      content: '';
      display: block;
      height: 2px;
      left: -3px;
      position: absolute;
      top: 3px;
      width: 14px;
    }
    .nano-charts-tooltip {
      background-color: rgba(0, 0, 0, 0.8);
      border: 1px solid #000;
      border-radius: 2px;
      color: #fafafa;
      font-family:Arial, Helvetica, sans-serif;
      font-size: 12px;
      left: 0;
      line-height: 16px;
      opacity: 0;
      padding: 16px;
      pointer-events: none;
      position: absolute;
      top: 0;
    }
    .nano-charts-tooltip.is-visible {
      opacity: 1;
    }
    .nano-charts-tooltip.is-visible:empty {
      opacity: 0;
    }
    .nano-charts-overlay {
      bottom: 30px;
      left: 70px;
      position: absolute;
      right: 70px;
      top: 30px;
    }
    .nano-charts-title {
      font-weight: 600;
      margin: 0 0 8px 0;
    }
    .nano-charts-item {
      margin: 0;
    }
    .nano-charts-badge {
      border-radius: 4px;
      display: inline-block;
      height: 8px;
      margin-right: 4px;
      width: 8px;
    }
  
  </style>
</div>
`;

exports[`Chart setData sets the data 1`] = `
<div
  style="height: 250px; position: relative;"
>
  <svg
    class="nano-charts"
    height="200"
    viewBox="0 0 1000 200"
    width="1000"
  >
    <g
      transform="translate(80, 40)"
    >
      <line
        stroke="#ececec"
        x1="0"
        x2="0"
        y1="-5"
        y2="125"
      />
      <line
        stroke="#ececec"
        x1="420"
        x2="420"
        y1="-5"
        y2="125"
      />
      <line
        stroke="#ececec"
        x1="840"
        x2="840"
        y1="-5"
        y2="125"
      />
      <line
        stroke="#ececec"
        x1="-20"
        x2="860"
        y1="120"
        y2="120"
      />
      <line
        stroke="#ececec"
        x1="-20"
        x2="860"
        y1="96"
        y2="96"
      />
      <line
        stroke="#ececec"
        x1="-20"
        x2="860"
        y1="72"
        y2="72"
      />
      <line
        stroke="#ececec"
        x1="-20"
        x2="860"
        y1="48"
        y2="48"
      />
      <line
        stroke="#ececec"
        x1="-20"
        x2="860"
        y1="24"
        y2="24"
      />
      <line
        stroke="#ececec"
        x1="-20"
        x2="860"
        y1="0"
        y2="0"
      />
      <path
        d="M 0 72 L 420 84 L 840 96"
        fill="transparent"
        stroke="violet"
        stroke-width="1"
      />
      <circle
        cx="0"
        cy="72"
        fill="violet"
        r="2"
      />
      <circle
        cx="420"
        cy="84"
        fill="violet"
        r="2"
      />
      <circle
        cx="840"
        cy="96"
        fill="violet"
        r="2"
      />
      <path
        d="M 0 72 L 420 84 L 840 96 L 840 72 L 0 72"
        fill="violet"
        stroke="transparent"
        stroke-width="1"
      />
      <path
        d="M 0 96 L 420 84 L 840 120"
        fill="transparent"
        stroke="#6a6a6a"
        stroke-width="1"
      />
      <circle
        cx="0"
        cy="96"
        fill="#6a6a6a"
        r="2"
      />
      <circle
        cx="420"
        cy="84"
        fill="#6a6a6a"
        r="2"
      />
      <circle
        cx="840"
        cy="120"
        fill="#6a6a6a"
        r="2"
      />
    </g>
    <g>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="middle"
        x="80"
        y="180"
      >
        Jan...
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="middle"
        x="500"
        y="180"
      >
        Feb...
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="middle"
        x="920"
        y="180"
      >
        March
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="end"
        x="40"
        y="160"
      >
        -20
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="end"
        x="40"
        y="136"
      >
        -10
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="end"
        x="40"
        y="112"
      >
        0
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="end"
        x="40"
        y="88"
      >
        10
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="end"
        x="40"
        y="64"
      >
        20
      </text>
      <text
        alignment-baseline="middle"
        font-size="12"
        text-anchor="end"
        x="40"
        y="40"
      >
        30
      </text>
    </g>
  </svg>
  <div
    class="nano-charts-tooltip"
  />
  <div
    class="nano-charts-overlay"
  />
  <div
    class="nano-charts-legend"
  >
    
        
    <span
      class="nano-charts-legend-item"
    >
      
          
      <span
        class="nano-charts-legend-item-badge"
        style="background-color: violet; color: violet;"
      />
      
          Metric A
        
    </span>
    
      
        
    <span
      class="nano-charts-legend-item"
    >
      
          
      <span
        class="nano-charts-legend-item-badge"
        style="background-color: #6a6a6a; color: #6a6a6a;"
      />
      
          Metric B
        
    </span>
    
      
  </div>
  <style>
    
    .nano-charts {
      font-family:Arial, Helvetica, sans-serif;
    }
    .nano-charts-label {
      line-height: 20;
    }
    .nano-charts-legend {
      background-color: #f9f9f9;
      display: flex;
      font-family:Arial, Helvetica, sans-serif;
      font-size: 12px;
      flex-wrap: wrap;
      justify-content: center;
      padding: 10px;
    }
    .nano-charts-legend-item {
      margin: 0 10px;
    }
    .nano-charts-legend-item-badge {
      border-radius: 4px;
      display: inline-block;
      height: 8px;
      margin-right: 6px;
      position: relative;
      width: 8px;
    }
    .nano-charts-legend-item-badge:before {
      background-color: currentColor;
      content: '';
      display: block;
      height: 2px;
      left: -3px;
      position: absolute;
      top: 3px;
      width: 14px;
    }
    .nano-charts-tooltip {
      background-color: rgba(0, 0, 0, 0.8);
      border: 1px solid #000;
      border-radius: 2px;
      color: #fafafa;
      font-family:Arial, Helvetica, sans-serif;
      font-size: 12px;
      left: 0;
      line-height: 16px;
      opacity: 0;
      padding: 16px;
      pointer-events: none;
      position: absolute;
      top: 0;
    }
    .nano-charts-tooltip.is-visible {
      opacity: 1;
    }
    .nano-charts-tooltip.is-visible:empty {
      opacity: 0;
    }
    .nano-charts-overlay {
      bottom: 30px;
      left: 70px;
      position: absolute;
      right: 70px;
      top: 30px;
    }
    .nano-charts-title {
      font-weight: 600;
      margin: 0 0 8px 0;
    }
    .nano-charts-item {
      margin: 0;
    }
    .nano-charts-badge {
      border-radius: 4px;
      display: inline-block;
      height: 8px;
      margin-right: 4px;
      width: 8px;
    }
  
  </style>
</div>
`;
